core/src/main/java/org/apache/iceberg/io/DeltaWriterFactory.java

```diff
@@ -0,0 +1,152 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.iceberg.io;
+
+import java.util.List;
+import org.apache.iceberg.ContentFileWriterFactory;
+import org.apache.iceberg.DataFile;
+import org.apache.iceberg.DeleteFile;
+import org.apache.iceberg.PartitionKey;
+import org.apache.iceberg.Schema;
+import org.apache.iceberg.deletes.PositionDelete;
+import org.apache.iceberg.relocated.com.google.common.base.Preconditions;
+import org.apache.iceberg.relocated.com.google.common.collect.ImmutableList;
+
+/**
+ * Factory to create {@link DeltaWriter}, which have few dependencies factories to create different kinds of writers.
+ */
+public interface DeltaWriterFactory<T> {
+
+  /**
+   * Create a factory to initialize the {@link DeltaWriter}.
+   */
+  DeltaWriter<T> createDeltaWriter(PartitionKey partitionKey, Context context);
+
+  /**
+   * Create a factory to initialize the {@link FileAppender}.
+   */
+  FileAppenderFactory<T> createFileAppenderFactory();
```

Comment:
Well,  we should only have one `FileAppenderFactory` for one `TaskWriter`,  each partition `DeltaWriter` could share the `FileAppenderFactory`.  It's better to move it out of this factory.

The following `createDataFileWriterFactory`,  `createEqualityDeleteWriterFactory` have the similar issues.